### ISSUE TITLE
Test Adapter runs explicit tests when TFS TestCaseFilter is used

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit3-vs-adapter#47]
See the original issue filed against  the 2.0 adapter: nunit/nunit-vs-adapter#71

**Comment by CharliePoole** [comment]
The 2.0 issue, nunit/nunit-vs-adapter#71, suggests two ways to deal with this problem:

1. Translate the TFS filter to an NUnit Filter.

2. Review the list of test names received from the filter after they are discovered by NUnit and remove any that are marked explicit.

Upon further review, (2) is not a solution. Tests need to be removed only if the filter does not explicitly select them. So, we have to parse the filter expression in either case... for (1) so we can generate the nunit filter and for (2) so we can retroactively correct the selection made by the filter in the first place. If we have to do so much work for (2) then I would lean toward (1) as the best approach.

That decided, however, another big choice looms!

In the V2 adapter, we implemented the use of MsTest-style filtering because it was the easiest approach. NUnit V2 didn't have it's own filtering language.

But in NUnit 3.0, we already have a filtering DSL, which is used by the console runner and is implemented as a service of the TestEngine. In fact, we could trivially transform an expression like "cat==Data || Priority == High" into a set of NUnit test filters.

We could also, with a bit more work, make a copy of our existing code to parse our DSL and change it to parse MsTest-style filters. As with our existing code, we would emit an nunit filter that honored the Explicit attribute. Of course, this would give us two DSLs to support and for users to remember.

Options:
1. Continue to use the MsTest filter language but parse it ourselves rather than leaving it to VS.
2. Replace it with our own DLS as used in the --where option of the console runner. This could be justified by the fact that the breaking change is being made prior to the 3.0 release.
3. Support both, which is obviously more work.
4. Support a hybrid of the two by allowing both MsTest-style expressions and NUnit expressions in the filter. This is probably the most work and might confuse users.

@OsirisTerje @rprouse What do you think?

**Comment by jnm2** [comment]
I can't get TestCategoryFilter to function at all in TFS 2015.1 build vNext with NUnit adapter 3.0.9.0. Is this related?

**Comment by CharliePoole** [comment]
@jnm2 No, this issue is a pretty well-known problem. I haven't tried vNext at all. If the problem persists, you may want to file another issue, showing details of what you are doing and what is happening (or not).

**Comment by CharliePoole** [comment]
Any rational way forward on this issue involves parsing the filter expression.

Parsing the filter expression requires deciding on the filtering DSL we want to support. The leading choices are our own existing DSL and the MsTest one.

Nobody is giving any feedback on the choice here, so I'm pulling the issue out of the milestone. Perhaps we can refer it to the general community for a future release.

**Comment by CharliePoole** [comment]
@OsirisTerje Nobody has ever responded to my request for comment on this, so I think it's up to us to decide what to do. Do you think we should continue to support the MsTest-style filter expressions here, require use of our own expressions as used in NUnit3-Console's --where option or support both?

Any option that supports MsTest-style filters requires us to parse them, of course. 

I'd like to either decide to move ahead with this in a certain direction or simply to say we are not planning to fix it.

**Comment by adamralph** [comment]
I raised https://github.com/nunit/nunit/issues/1664 and was directed here.

Can some please confirm whether this issue also represents the problem purely in the VS Test Explorer, regardless of TFS?

I'm not using TFS, but when I run "all tests" in the VS Test Explorer, test marked `Explicit` are being run.

**Comment by rprouse** [comment]
@adamralph yes this is an issue with just VS Test Explorer also.

**Comment by adamralph** [comment]
Thanks @rprouse.

That's not at all obvious from the title and description. Would it be worth refining, to make it easier to find for anyone else having the same problem? This is the reason I couldn't find it when I searched, and why I raised https://github.com/nunit/nunit/issues/1664 instead.

**Comment by dnperfors** [comment]
@CharliePoole shouldn't MsTest-style filters be supported in the vs-adapter anyways, so that it correctly works from within Visual Studio and TFS-Build?
I hope this issue can be solved soon, since we get quite a lot of problems on our buildserver (which is effectively using the VS Test Explorer)

**Comment by CharliePoole** [comment]
@rprouse Are you sure @adamralph 's problem is the same? In nunit/nunit#1664 he talks about using Run All from the IDE and doesn't mention any use of filters. Obviously it is an adapter issue, but I'm not so sure it's the same as this one.

@adamralph Can you comment on this? If we can't draw a connection with this one - i.e. if no filters are involved - then we will need to confirm it separately. It's not a problem I have ever seen. Specifically, the NUnit 3 Test Demo project contains tests marked explicit and they are handled as expected.

**Comment by ChrisMaddock** [comment]
I believe @adamralph's problem may also be with the v2 adapter rather than this one, I had a quick dig earlier. 

@adamralph - is that correct, the issue was with the NServiceBus tests, running NUnit v2.6.4?

**Comment by CharliePoole** [comment]
@dnfperfors There's no requirement of supporting MsTest style filters in order to run under TFS and the IDE. The user supplies the filter as an expression and it's up to the adapter to interpret it.

The solution to this problem requires a significant amount of work - but rather interesting work if you like parsing. It can go in two completely different ways, so I asked for feedback three months ago. Yours is the first reply.

Our options are:

1. Support MsTest-format filters. If we do that, users will have to use one style of filter if working in the adapter and a different one when using the console or gui runners. We will have to write a parser in order to resolve the Explicit problem, probably basing it on our existing parser.

2. Support our own filters. If we do that, users will use one filter expression for all NUnit runners but a different one if they are working with MsTest. We will reuse our existing parser code. The Explicit problem will be resolved as a side effect. Existing users of MsTest-style filters will need to change to use NUnit Test Selection Language.

3. Support both kinds of filters. In this case there are sub-options.

  3.1 Leave the MsTest-style filters broken with respect to explicit. We would set up some way for the user to tell us which type of filter they were using, probably by requiring some special prefix to trigger our own.

  3.2 Fix the MsTest-style filters as well. This requires all the same work as both options 1 and 2. This is probably the "ideal" solution, but only if we know that folks want both options. That's why we ask for feedback!

Permit me to briefly rant...

I've been working on NUnit for about 14 years now. When asking for feedback, I have often been told "Just do what you think best - you're the one working on it." In many cases, I have taken this to heart because at least it gives a decision and we can move forward.

That approach doesn't work for everything, however. Sometimes I __absolutely__ need user input, so I ask for it. This is such a case. Because we are talking about a feature that I personally __never__ use, I can't have an intelligent opinion about it. All I can do is outline the options and the pros and cons that I see. Others have to respond to that if they want the thing to move ahead. If not enough people care strongly enough to give some feedback, then it doesn't move. There are other things in NUnit that I feel strongly about myself, so I focus on them.

End of rant...

Currently this issue is in the 3.5 milestone but with nobody assigned. That''s a bit of an anomaly. I put it there in the hopes of drawing attention to the fact that it is blocked. I should have marked it blocked, which I just did, for the reasons explained above.

**Comment by CharliePoole** [comment]
@ChrisMaddock Good on you for noticing that. We only have a one line problem description from @adamralph so let's wait for more information before replicating this anywhere else.

**Comment by adamralph** [comment]
I'll try and push a repro solution to GitHub some time tomorrow.

**Comment by rprouse** [comment]
@CharliePoole I had assumed that Run All was passing an MSTest style filter with the list of all tests to run and that is why the explicit tests are being run, but I haven't debugged it, so I might be mistaken. I am seeing the same behaviour in the .NET Core adapter.

**Comment by CharliePoole** [comment]
@rprouse If it's doing that, it's new behavior on VS's part and violates their documentation, which explicitly states that the kind of filter we use is up to us. I'll check in the debugger with VS 2015.

**Comment by dnperfors** [comment]
@CharliePoole, thanks for the explanation. I don't think we use the MsTest-style filters and even if we do, it would be no problem to change them to the NUnit-style filters. So I would opt for the quicker solution of at least start supporting the NUnit filters in 3.5 and, if necessary, include support for MsTest-style filters later.

**Comment by CharliePoole** [comment]
@dnperfors Thanks for the info. Yes, we could clearly do this incrementally, in which case starting with the easier part is a good idea.

**Comment by CharliePoole** [comment]
I've verified that @adamralph 's problem doesn't happen on the NUnit 3 adapter. When using Run All, tests marked Explicit show up as Not Run as you would expect. I'll next check it with the NUnit (2) Adapter.

**Comment by adamralph** [comment]
I guess the VS test explorer thread has moved to https://github.com/nunit/nunit-vs-adapter/issues/125 so I'll chime in there.

**Comment by CharliePoole** [comment]
@adamralph Always assuming that's where it belongs... :-)

**Comment by benbenbenbenbenben** [comment]
On line 187 of NUnit3Executor.cs is says:

` var loadResult = _activeRunner.Explore(TestFilter.Empty);`

Can't we use the runstate attribute in loadResult to determine what tests are Explicit and then remove any whose fullname **is** a qualified name in the list generated from mstest/vstest test filter discover but **is not** specified in the mstest/vstest filter?

**Comment by richardjm** [comment]
Just thought I'd let you know this is now an issue with the new "Live unit testing" feature of VS 2017. Which merrily runs explicit tests.

**Comment by CharliePoole** [comment]
Good to know. It's clearly a side effect but will influence how we fix it. It's unlikely that Microsoft's feature would ever use the NUnit selection language, so we will most likely need to parse their expressions.

**Comment by CharliePoole** [comment]
@nunit/core-team This single-digit issue has been around for almost two years now. I made a proposal for (unfortunately) two alternative ways to fix it about a year ago. I could easily create the grammar to parse either of the two formats, just as we parse our own TSL. I think this is primarily up to @rprouse and @OsirisTerje to make a decision, with input from others of course. The questions that need answering are:

1. Should we support TSL as a filtering mechanism in the adapters, OR
2. Should we support TFS filtering language, OR
3. Should we (somehow) support both.

For (3) we would have to define whether the user uses a different attribute for each language or if we will somehow auto-detect which is being used or if we will create a syntax that allows combining both in one expression. All are reasonably feasible.

**Comment by jnm2** [comment]
(TFS is a bit of a misnomer. Technically it's the VSTest console `/TestCaseFilter` argument, present in every VS install and used by Cake scripts.)

The only upside I see to supporting a common standard for the VSTest TestCaseFilter is that you may have both MSTest and NUnit tests being run at the same time. That was certainly the case for a long period of time as I incrementally converted us at work to NUnit.

You would expect all adapters to behave intuitively and give a best-effort interpretation of the filter; anything else is inflexible and leaves you having to convert whole assemblies at a time and maintain two separate test runs separated by assembly. The ignore/explicit thing is enough of a painful recent memory that I think we should support the VSTest console `/TestCaseFilter` filter format when passed in via  `/TestCaseFilter`. I'm not sure what supporting both would look like but I'm skeptical that it would be valuable. A separate off-by-default switch could change the interpretation of `/TestCaseFilter` from the VSTest format to the NUnit format if we hear from folks wanting to use it that way.

**Comment by jnm2** [comment]
By best-effort, I would expect Ignore and Explicit to still be effective even while interpreting the `/TestCaseFilter` argument according to the VSTest format.

**Comment by CharliePoole** [comment]
The entire purpose of doing this issue is to make Explicit work right, so I agree that it should work when we are done!

This has nothing to do with ignored tests, which work perfectly well with filtering. That's because NUnit had no special semantics around ignored tests.

I see your point about the VS syntax. It's what we currently support, albeit with this particular bug. Our own TSL has more features, but I don't see anyone asking for them. Folks wanting those features can use the console runner.

Supporting both would be like supporting foreign words in English. The two syntaxes ate very similar.

Ciao, Charlie

**Comment by CharliePoole** [comment]
@jnm2 There is nothing in the VS syntax that makes it more difficult to handle Explicit tests. The problem is that VS is not aware of Explicit. If we parse the expression and create a proper NUnit filter, then Explicit will be taken care of by NUnit itself.

**Comment by jnm2** [comment]
All sounds good to me.

**Comment by OsirisTerje** [comment]
I've been trying to get a consistent view of these issues, but it is hard.  There are several issues connected with Explicit (and also some with Ignore, but that can be handled in #241, see my comment there).  
First VS:
Explicit alone works with VS,( but give an error message in VSTS).  
Explicit with a filter in VS (which is NOT the same filter as TFS, and which are not given to us, but instead calls the other execute method with an array of testcases) is just behaving very strange.
 
My test code is as:
` public class Class1
    {
        [Category("C")]
        [Category("B")]
        [Category("A")]
        [Test]
        public void TestIt()
        {
            Assert.Pass();
        }

        [Category("B")]
        [Explicit]
        [Test]
        public void TestItExplicit()
        {
            Assert.Pass();
        }

        [Category("C")]
        [Ignore("because")]
        [Test]
        public void TestItIgnored()
        {
            Assert.Pass();
        }`

The results in VS with no filters are as expected:
  
![image](https://user-images.githubusercontent.com/203432/28437958-ccecfaa8-6d9c-11e7-9b78-8fc67f08096d.png)

Adding a category/trait filter which should result in only the first test being shown:
![image](https://user-images.githubusercontent.com/203432/28438022-fc878ac6-6d9c-11e7-80a6-c916c1ac9dd9.png)

This is just weird.

Moving to VSTS:
With no filters:
![image](https://user-images.githubusercontent.com/203432/28438209-a4c1fa82-6d9d-11e7-8b41-d5d86048095b.png)

The explicit test is reported as failed.
Also note that the build is reported as success, but should have been reported as partially success if it really was a failed test.  
This corresponds to a run with vstest.console:
![image](https://user-images.githubusercontent.com/203432/28438289-edc1f2b4-6d9d-11e7-94bc-4ba7c6ef2957.png)

This is without filters, and obviously this is not working.  We have issues with both the status we return and with the error/output messages left.  

(Need to stop here now, will add filter results tomorrow)

**Comment by CharliePoole** [comment]
I wrote this issue to deal with the problem that Explicit tests run under a TFS filter execute even when not explicitly selected. While I realize there are many other Explicit-ish problems, I think life is usually simpler when we solve the problems we understand first.

The cause of the TFS fltering issue with explicit is that we don't implement the filtering ourselves but instead ask the TFS filter to apply itself to the tests and tell us what to run. Naturally, the Microsoft code knows nothing about ExplicitAttribute or explicit selection in NUnit. We receive a list of names and execute them. Various stopgaps were proposed over time - probably listed above - but none of them work. In general, if the list we are given includes Explicit tests, we can't tell whether they should be run without analyzing the filter.

The clear solution to this problem is that we should translate the filter syntax into an NUnit test filter, which (if we do it correctly) will simply work because NUnit understands Explicit.

I've never tested using trait filters in VS, but I think that is a different matter. In the case of TFS filters, we actually have the filter expression in hand but we choose to tell VS to do the filtering. In the case of traits filtering, I don't think we ever know how the list of tests to run was created, so I don't think we can fix it. It still may be useful to create an issue for this, just to document the problem. I don't think that has ever been done.

I'd definitely like your opinion on which syntax to support for the so-called TFS filter.

**Comment by benbenbenbenbenben** [comment]
@OsirisTerje Because the vstest TFS add-in looks for a filename ending vstest.exe we wrote a wrapping executable to the Exclusive problem on our TFS builds.

All our environment team has to do is put our path in place when choosing the "specify vstest executable" option and we have a properly functioning nunit filter.

I can't share code but I can tell everyone how it works;

1. It's an executable whose name ends .vstest.exe
2. It intercepts the filter string from the configuration page in TFS (which is where we do or do not name explicit tests)
3. Examines the test assemblies to see which tests are marked Exclusive
4. It then converts the filter string + exclusive test information to the "actual" list-of-test filter vstest.exe would use
5. Writes said list-of-test filter to file (so we can pass: /TestCaseFilter:File=FILTER to the original vstest.exe)
6. Runs vstest.exe passing stdout through untouched and thereby appearing to be vstest.exe

If I remember correctly we used TinyPG to make the engine to parse filters properly.

Complicated it may be... but it worked.

**Comment by jnm2** [comment]
I finally do have an opinion on the direction now. I think there's great value in being consistent with the `dotnet` CLI, `dotnet test` and `dotnet vstest`:

> ## Filter option details
> 
> `--filter <EXPRESSION>`
> 
> `<Expression>` has the format `<property><operator><value>[|&<Expression>]`.
> 
> `<property>` is an attribute of the `Test Case`. The following are the properties supported by popular unit test frameworks:
> 
> | Test Framework | Supported properties                                                                                      |
> | :------------: | --------------------------------------------------------------------------------------------------------- |
> | MSTest         | <ul><li>FullyQualifiedName</li><li>Name</li><li>ClassName</li><li>Priority</li><li>TestCategory</li></ul> |
> | Xunit          | <ul><li>FullyQualifiedName</li><li>DisplayName</li><li>Traits</li></ul>                                   |

(And the rest of https://docs.microsoft.com/dotnet/core/tools/dotnet-test#filter-option-details)

**When** (not if 😁) we package NUnit Console as `dotnet nunit`, that is where we should definitely keep our own test filter language.

**Comment by CharliePoole** [comment]
Our code at https://github.com/nunit/nunit3-vs-adapter/blob/master/src/NUnitTestAdapter/TFSTestFilter.cs is supposed to be supporting a whole list of things. If it isn't, we must not have tests for it. 😈

**Comment by jnm2** [comment]
Since VSTest now allows you to make up your own keywords like `TestCategory != X & Nonsense` (https://github.com/Microsoft/vstest/pull/845), how about the following?

- We add a trait named Explicit to explicit test cases so they can be grouped and filtered in Test Explorer.
- We do the built-in command-line filtering but tweak it:
  `TestCategory = Navigation`—no explicit tests
  `TestCategory = Navigation & TestCategory != Explicit`—no explicit tests
  `TestCategory = Navigation & Explicit`—only explicit tests
  `TestCategory = Navigation & TestCategory = Explicit`—only explicit tests
  `TestCategory = Navigation & (TestCategory = Explicit | TestCategory != Explicit)`—both explicit and non-explicit tests

I wonder if someone from the VSTest team would have suggestions or knowledge of the future for how to make these better.

Optional:
- Since the Test Explorer UI is not keyed in to the concept of explicit selection the way ReSharper is, we can't match 1:1. Currently explicit tests are included for any selection except for Run All.
  What if we change it so that we always exclude explicit tests from the Test Explorer selection unless ONLY explicit tests have been selected?

**Comment by CharliePoole** [comment]
We've had two clear paths to fixing this for a long time. Either one would work just as NUnit users have come to expect, i.e. like NUnit console runner does. I asked for input way back when because the two approaches gave totally different interfaces to the user - VS test selection language versus NUnit test selection language. If we could have made a decision then, one of those options would have been implenented long ago!

The approach of making the user specify how to handle explicit seems non-friendly and not particularly easier than simply parsing the selection expression. The problem comes from the fact that we are allowing VS to decide what the expression means rather than deciding ourselves. Or, probably VS folks would say that the problem is that our semantics do not match their expectations. 😄 

Not sure what you are saying about "not matching 1:1" - If you mean matching what NUnit is supposed to do with explicit, it seems perfectly feasible to me.

**Comment by jnm2** [comment]
No, I mean that I think we should respect the VSTest format so that tools that work with VSTest can pass filters to VSTest and expect them to work.

**Comment by jnm2** [comment]
People who want to use our filter language from the command line will not be running vstest.

**Comment by CharliePoole** [comment]
Yes, I understood that. You already said that was the option (of the two listed) you preferred. I'm talking about this new, third option you described, where we continue to allow VS to interpret the filters but create special categories for Explicit.

Perhaps I didn't make myself clear earlier...

What we do now is tell VS to apply the filter and give us a list of the tests to run. By doing that, it's the VS semantics about Explicit that get applied - or rather the absence of semantics.

What I suggested we do is directly translate the VS-provided filter to an NUnit filter. That will automatically apply NUnit semantics.

**Comment by jnm2** [comment]
My apologies, I understand now. I favor offloading as much of the work as possible to VS's MatchTestCases so that brand new VSTest filter operators work without duplicate work from us. If we expose Explicit as a trait, we won't even have to do that much modification.

With my proposal above, the only modification to the default behavior that we'd have to make is to notice when the filter string does not contain the trait `Explicit` and therefore to modify it to `({originalFilter}) & TestCategory != Explicit` before passing it to VS.

**Comment by jnm2** [comment]
My apologies, I understand now. I favor offloading as much of the work as possible to VS's MatchTestCases so that brand new VSTest filter operators work without duplicate work from us. If we expose Explicit as a trait, we won't even have to do that much modification.

With my proposal above, the only modification to the default behavior that we'd have to make is to notice when the filter string does not contain the trait `Explicit` and therefore to additionally check that before asking VSTest about each test case.

**Comment by CharliePoole** [comment]
Let's call this option 3 then.

I don't like it because it offloads the problem to the user rather than solving it.

Also, "asking VStest about" the individual tests is unclear and AFAICS unsupported. If it were possible to ask which tests were explicitly selected, the issue wouldn't exist. The VA TestCase only contains the info we give to VS in the first place, during discovery.

**Comment by jnm2** [comment]
> Also, "asking VStest about" the individual tests is unclear and AFAICS unsupported. If it were possible to ask which tests were explicitly selected, the issue wouldn't exist. 

I wasn't specific enough. Here I'm talking about the command line filter string, not the Test Explorer UI, and "asking VSTest about" means precisely the bool-returning `ITestCaseFilterExpression.MatchTestCase`:

https://github.com/nunit/nunit3-vs-adapter/blob/323b598172c6fa0f47955890e2930b6951e80817/src/NUnitTestAdapter/TfsTestFilter.cs#L108

We delegate the parsing and evaluation of the filter to VSTest and I see value in keeping it that wasy as much as possible.

> I don't like it because it offloads the problem to the user rather than solving it.

In what way does it offload the problem to the user?

**Comment by CharliePoole** [comment]
It's clear we are talking about command-line filters. It's in the title of the issue, which I wrote.

The line in question asks VS to apply its filter, giving us a list of VS test cases to run. We translate those back to the NUnit test cases and create a filter to run them. While we know which test cases have an Explicit attribute, we __don't__ know which of them were explicitly selected by the user, rather than indirectly by having a parent selected. Hence, we run them all. 

We could instead skip explicit test cases, but that would mean they could __never__ be run from the adapter.

The extra info we __need__ about test cases is how they were or were not selected by the user. We can only get that information if we examine the text of the filter itself, which is available to us. I considered an adhoc approach that does not involve parsing, but it appears that parsing - which is pretty trivial for anyone who knows how - is the only way to get completely accurate information. That makes sense: you have to understand the filter in order to know __why__ a particular test is selected. Most likely, there is some internal VS structure that we could examine, which would give us the information even more easily, but we don't have access to such a structure.

All this comes from the fact that we are an __adapter__ and not an __extension__ to VS. We have to work with the information that is given to us, based on what the VS Test team imagined frameworks would need to know. It's probably good enough for any frameworks created __after__ the adapter API but it doesn't fit well with pre-existing frameworks, with the obvious exception of MSTest.

You are putting the problem in the hands of the user because you are requiring them to use the Explicit "trait" in order to express what they want. As far as I can see, it still doesn't allow them to express everything.

Consider this...

```C#
[Category("X")]
public class SomeClass
{
    [Explicit]
    public void SomeTest() { }

    public void SomeOtherTest() { }
}
```

IF my filter is "TestCategory = X", I am told to run `SomeTest` and `SomeOtherTest` - but I should not run `SomeTest` so that's incorrect.

IF my filter is "TestCategory=X | Name=SomeOtherTest", I am told to run both tests again - but this time it's correct.

The difference is only available by inspecting the filter, which we do not now do. Once we have analyzed (parsed) the filter, we may as well just create an NUnit filter out of it rather than using the less efficient approach of enumerating all the tests. Note that some filter expressions may create a list of thousands of tests!

If my filter is "TestCategory =

**Comment by jnm2** [comment]
> While we know which test cases have an Explicit attribute, we don't know which of them were explicitly selected by the user, rather than indirectly by having a parent selected. Hence, we run them all.

I know how you can select a parent in the UI, but I don't know how you can select a parent from the command line because the internal API applies the filter to only test cases (leaf nodes).

> The difference is only available by inspecting the filter, which we do not now do.

I agree.

> Once we have analyzed (parsed) the filter, we may as well just create an NUnit filter out of it rather than using the less efficient approach of enumerating all the tests. Note that some filter expressions may create a list of thousands of tests!

I don't think we need to parse it. If the single word `Explicit` appears in the filter, our job is done. (With my proposal above.) If it doesn't we need to simulate adding `(...) & TestCategory != Explicit` to the filter.

**Comment by CharliePoole** [comment]
Unlike "native" NUnit, the traits which we create for each test case do include the categories and properties of the class that contains them. That's part of being an adapter. 😄 

The use of the word "explicit" in the filter doesn't tell you how it's used. Unless what you are saying is that the user only has two options: include all explicit tests or exclude them. But that doesn't match the full flexibility of NUnit filters, which exclude all explicit tests __except__ those that are explicitly selected.

**Comment by CharliePoole** [comment]
Consider a test with `[Category("Navigation"), Explicit]`. User passes `TestCategory = Navigation` as in your first example. We should __not__ filter out the test.

All this assumes, on my part, that we want NUnit to work as it normally does.

**Comment by jnm2** [comment]
@CharliePoole I believe we should filter out that test on the principle of making the 99% case easy and the 1% case not too difficult (`TestCategory = Navigation & Explicit` if you want that test case).

**Comment by jnm2** [comment]
@CharliePoole I believe we should filter out that test on the principle of making the 99% case easy and the 1% case not too difficult (`TestCategory = Navigation & Explicit` if you want that test case).

Most of the time `TestCategory = Navigation` means, run my Navigation section of tests. Do we really want every build script to have to opt out of explicit tests? How often are explicit tests run? Only during debugging and vanishingly rarely from the VSTest.Console.exe CLI.

**Comment by CharliePoole** [comment]
Except I proposed an option to make it work 100% of the time, so it is an option.

Your arguments are not all wrong if you want to talk about how Explicit should work in NUnit and if we want to make a breaking change to it. But when it comes to the adapter I think the behavior should match NUnit native behavior unless VS makes that impossible. It does in some cases, but not here.

Another possibility would be to mark this as won't fix and allow an nunit filter to be specified. Then we could say "If you want the NUnit behavior, use an NUnit filter."

@OsirisTerje OK think you need to make a decision.

**Comment by jnm2** [comment]
NUnit behavior for the filter `cat == Navigation` would be to skip all explicit tests that otherwise matched so we agree that far, right? That's the most important to me so that build scripts can use `/TestCaseFilter:"TestCategory = Navigation"` to also skip all explicit tests that otherwise matched.

**Comment by jnm2** [comment]
Oh my. I'm struggling to visualize this. I'm using `--where "cat != Integration"` and never realized it was running my explicit tests. So if Category is on the fixture and `[Explicit]` is on the method, the explicit test will be skipped, but if Category and Explicit are both on the method, it won't be?

**Comment by jnm2** [comment]
With such a simplified understanding as I have of NUnit's own filtering, I prefer to step back from the design discussion. Happy to implement if no one else was hoping get to do it!

**Comment by OsirisTerje** [comment]
The major problem is the title of this item, which I see we have in our builds too, that a TFS build, which uses the VSTest task, which uses the VSTest.console runner, which uses the adapter to run NUnit, runs explicit tests when the users don't expect them to be run. 

1) We need to stop this behaviour.   The default behaviour should be to **not** run explicit tests, if not explicitly being told to do so. 

2) Then we need to establish some ways that the user **can** run explicit tests.    I am not so concerned whether we follow the exact NUnit behaviour, I am not even quite sure about the details of that, Joe is not so sure about it either.  I know that we can't tell the users to replace the TFS/VSTest filter syntax with NUnit's, no way people will change all their build definitions for that (if that was the suggestion).   Earlier VSTest didn't allow us to use our own keywords, now they do (since  https://github.com/Microsoft/vstest/pull/845, which @jnm2  found).  So now we can do that, and thus extend the existing VSTest syntax with our own keyword to run explicit tests, without breaking VSTest.   I think that is a good idea. 

I might have misunderstood things here, but I try to rephrase it:   
Option a)  Replace with NUnit filtering syntax, and allow full NUnit filtering semantics
Option b)  Stay with the VSTest filtering syntax, but extend it, and make it match the NUnit semantics as far as we can get it.

I believe, that if we can get the exact NUnit filtering semantic to work, that would be preferably, option b).   I looked up the filtering documentation (on the explicit attribute) I could find, but what @CharliePoole said above, I can't find it in the docs.   Do we have this clearly documented?

Now, should we parse the expression ourselves fully , in option b), or should we rely on VSTest to do that for us, as we do now.   The VSTest filtering language (DSL) is rather simple, so I don't think it will take much to implement that.  But, I am (just a bit) concerned about us duplicating stuff that is not really ours (by design I mean)

Ok, just found (more of) the filter language specification here:  https://github.com/nunit/docs/wiki/Test-Selection-Language ,  it was under the NUnit.Console.  

About option a).  I don't see how we can replace the VSTest syntax with NUnit syntax at all.  VSTest will not accept our syntax, as it will try to parse it.  Did you mean this @CharliePoole , or did you mean something else?  

Even if we parse the VSTest filter, we don't know about the explicit tests.  As @CharliePoole said above, we don't know how the user has selected tests.  
If we go for the very strict option, like R# does, we could see if all tests that are in a selected list of tests are explicit (since we know the attributes), if they are, we run them, if not, we don't.    That would be an either/or approach, and would not require any new filter syntax.  

If we go for a possibility of a mixed approach, - and I am not sure where that would be used, and why anyone really would want that ( they set the explicit for a reason!), we need to sort of the filter discussion here.

**Comment by OsirisTerje** [comment]
About the https://github.com/nunit/nunit-vs-adapter/issues/71 issue:  The NUnit2 adapter has the extra bug that it doesn't run explicit tests when there is no filter, but runs them when there is a filter.  The NUnit3 adapter runs them in both cases.

**Comment by jnm2** [comment]
@OsirisTerje  I'm not sure the NUnit3 adapter runs them in both cases. Explicit tests were not running until I added `TestCategory != Navigation`.

https://github.com/nunit/nunit3-vs-adapter/blob/b4703ae4c8e10db76fb226c837c3a42199a32574/build.cake#L181-L183

**Comment by OsirisTerje** [comment]
Taking this from one of @CharliePoole 's comment on nunit/nunit-vs-adapter#71 :  _"Directly selected means either by name or by inclusion through a category. Excluding a category does not cause the explicit test to be run._ ".     

The strict mode above would not run "by inclusion through a category".     

In the Test Explorer, it will always run explicit tests by a selection, not only by a category, but by anything, class, namespace, project, property, trait or category, simply because these are all selections.
In the Test Explorer, this is annoying, but not something we can do anything about, since the filtering mechanism here is different from the vstest.console.  I don't think we can extend that, other than faking a trait. 
The inclusion on the builds, which are using the vstest.console, however, are more serious, and that is where we need to stop them being run.

**Comment by OsirisTerje** [comment]
@jnm2  Right you are, my slip.  I just saw it said 5 tests where I expected 4, didn't see one of them marked as "Not run".  Adding the screenshots below so we have this documented:
![image](https://user-images.githubusercontent.com/203432/36395865-a0a31190-15bb-11e8-83a8-03019350c479.png)

And with a filter set (which is bogus btw):
![image](https://user-images.githubusercontent.com/203432/36395878-b518eab4-15bb-11e8-836f-c400344f4c88.png)

**Comment by OsirisTerje** [comment]
On the 71 issue, @bussnik  has the workaround of adding Explicit as a category, and use that for achieving the goal.  We could do an automatic addition of Explicit as a trait too, and thus make it easier for users to use this workaround.

**Comment by rprouse** [comment]
Personally, I like the idea of adding Explicit as a category for the adapter. I solves the command line problem and also gives you a nice grouping in the UI for running those tests.

**Comment by jnm2** [comment]
Assuming we don't do anything special to favor skipping explicit tests, we should call out that all build scripts will want to use `& TestCategory != Explicit` so that people are aware of the default behavior.

Right now build scripts with no filter do not run explicit tests and build scripts with a filter do. Should we unify that so that we have consistent behavior whether or not a filter is specified?

We've agreed that it doesn't make sense for build scripts to run them. The other place this filter can be specified is e.g. `dotnet test --filter "TestCategory = Navigation"`.

**Comment by OsirisTerje** [comment]
I really don't like that we run explicit tests just because that a filter is being set, so I am pretty sure the default behaviour should be that no explicit tests are being run, filter or no filter, if not explicitly set. 
Using the special category Explicit can be a good way of stating that we want them run.

**Comment by jnm2** [comment]
If you decide to unify it, and have explicit tests skipped in both scenarios, how would you like the user to opt in?

When we get to the implementation phase, integration tests would be nice.

**Comment by OsirisTerje** [comment]
1.  By default:  All explicit tests are skipped, regardless how you run. This also makes it consistent with NUnit default behaviour.  Also consistent with @CharliePoole's suggestion https://github.com/nunit/nunit-vs-adapter/issues/71#issuecomment-122292674  , see Fix 2.
2.  Adding TestCategory = Explicit will run only Explicit tests.
3.  Mixing in with other categories will allow them to run too.

Options:
4.  TestCategory=Explicit can be shortened to just Explicit

Note that the only thing we don't do then is to allow an explicit test to be run because it is part of a test fixture marked with another selected category.  I don't understand the logic of this though (my understanding of this might even be wrong).
 
5. We need to decide how to handle:
a)  [Category("Navigation"), Explicit]    
@CharliePoole says this is to be run, as per NUnit semantics.   That both makes sense to me, but it doesn't when this is part of a hierarchy.    However, I don't want to break NUnit semantics. 
I would also like to see the feasibility of this from @CharliePoole : "What I suggested we do is directly translate the VS-provided filter to an NUnit filter. That will automatically apply NUnit semantics."  I don't see how we add the Explicit user choice, which has to be part of the filter.    
But we can do 1-4 first, and then 5 later.

**Comment by CharliePoole** [comment]
Looking back before moving ahead (in the next comment)...

@OsirisTerje You summarized my options as follows:

> I might have misunderstood things here, but I try to rephrase it:
Option a) Replace with NUnit filtering syntax, and allow full NUnit filtering semantics
Option b) Stay with the VSTest filtering syntax, but extend it, and make it match the NUnit semantics as far as we can get it.

That's not quite it. In my comment back in April, 2016 🙄 I suggested four things...
1. Support the VS syntax by parsing it ourselves
2. Support our own syntax in the same way
3. Support both
4. Support a hybrid

By "support" I had no intention to suggest that vstest-console would accept our own syntax. I assumed we would have to provide it through the `.runsettings` file.

My own preference these days is 1 + 2. Command-line filters would have to use the VS format. Adding our own syntax as a runsetting would be an extra.

**Comment by CharliePoole** [comment]
@OsirisTerje Why do we need a command-line switch to run explicit tests? We don't have a way to do this in the console runner and the framework has no way to overrule explicit except by specifically selecting the test. Selection can be by name, by category, etc.

In any case, if you feel we need this new feature, it's orthogonal to the purpose of this issue, which is to __stop__ running explicit tests when they are not selected.

**Comment by CharliePoole** [comment]
@OsirisTerje I'm not understanding the "extra" bug in the V2 adapter. You state that explicit tests are skipped when no filter is used. That's correct behavior. Is there some other bug?

**Comment by CharliePoole** [comment]
@OsirisTerje You wrote

> Note that the only thing we don't do then is to allow an explicit test to be run because it is part of a test fixture marked with another selected category. I don't understand the logic of this though (my understanding of this might even be wrong).

I think you mean something like

```C#
[Category("X")]
public class Fixture()
{
   [Explicit, Test]
   public void TheTest() { }
}

When running category "X", `TheTest` will not be executed because it is not explicitly selected. Explicit selection means you have asked to run the test by name or by a category that appears on it. (Non-explicit selection means you have asked to run a test suite that contains the test.)

**Comment by jnm2** [comment]
I wish category selection was in the same genre as parent selection so that it would never explicitly select a test case. I can't imagine thinking of a category as a way to single out an explicit test. But that's determined by the framework.

**Comment by CharliePoole** [comment]
I think that one thing being missed in the discussion is the __purpose__ of Explicit. It was designed for use on tests that one does not normally want to run as part of the CI build. In rare cases, one may want to override this behavior, which is possible through explicitly calling out the test to be run. 

A situation where you want to run __all__ your non-CI tests in the CI build seems pretty far-fetched. For example, I use `[Explicit]` myself...

1. When the test does UI and I am running headless in CI
2. When I want to actually see the result of a failure

I would never want such tests to run in CI.

**Comment by OsirisTerje** [comment]
@CharliePoole Thanks, the comments above really cleared it out for me.  :-)  

Then we have the default behavior cleared out, (stop running explicit tests) and we should get that fixed. (My pt.1. above).  We should have this in with 3.10.

To allow people to run explicit tests, I think the parsing to NUnit filter and letting the engine do the rest, is the best way to go, since, if I understand the engine correctly, it already has that functionality.   But, it takes some more time to implement that.  I'll add a separate issue for implementing that post 3.10, for 3.11. 
Until this is fixed, we keep the behavior in VS mode, that is that selected tests are run, regardless of how they are selected.  Since that is a separate Runtests interface method, we know when this is being selected. 


We can then just auto-add Explicit as a category, as was suggested above, which would also provide grouping in the UI as @rprouse points out, and it "unblocks us from blocking" when we implement the default behavior.  And if someone really wants to run them in CI, they can.   Even if I agree with @CharliePoole , can't really see why.  But at least we don't block anyone.   We do this in 3.10.  



@CharliePoole The bug in NUnit2 adapter is that it runs explicit tests when any kind of test filter is added, even if it doesn't points to the explicit test. So based on our earlier chat about that, we should fix it there too.

**Comment by CharliePoole** [comment]
@OsirisTerje That clears things up for me. We should continue the chat we started offline as well. That will help me figure out where or if I can help you.

**Comment by CharliePoole** [comment]
@OsirisTerje Is there a schedule for 3.10?

**Comment by jnm2** [comment]
It's not impossible that people might use `dotnet test --filter X` (which gives /TestCaseFilter:` to VSTest.Console.exe) for non-CI purposes, so they might appreciate having a way to run explicit tests.

**Comment by OsirisTerje** [comment]
@CharliePoole Yes, it is scheduled for release by the end of the month, that means, I hope to have it out by the end of next week.

**Comment by jnm2** [comment]
If I understood right, @OsirisTerje slated this for 3.10:

- [ ] Add an `Explicit` trait to test cases with `[Explicit]`
      <ul><li>This trait will be inherited from parent nodes the same way that traits created from categories do (in the translation to VS's leaf-only test cases)</li></ul>

- [ ] Skip all explicit tests when using a TestCaseFilter
      <ul><li>There is intended to be no way to run explicit tests while using TestCaseFilter in 3.10, to be addressed in the next release</li></ul>

And this for 3.next:

- [ ] Parsing TestCaseFilters to enable certain patterns to explicitly select test cases

@OsirisTerje Any details missing for the 3.10 work?

**Comment by OsirisTerje** [comment]
Seems correct , thanks!

**Comment by jnm2** [comment]
Should we start a new issue for 3.11 to discuss ways to run explicit tests from the command line?

**Comment by OsirisTerje** [comment]
@jnm2  Please do

**Comment by jnm2** [comment]
Done! https://github.com/nunit/nunit3-vs-adapter/issues/473
